star-sys/STAR/source/MmapArray.cpp

```diff
@@ -0,0 +1,85 @@
+# include "MmapArray.h"
+
+#include <string>
+#include <iostream>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+
+#define PAGE ((size_t) (1<<12))
+
+size_t round_up_to_page(size_t v) {
+    return (v + (PAGE - 1)) & ~(PAGE-1);
+}
+
+// mmap `filename` as a mostly read-only map, allowing for limited writing before an after the file.
+// Maps at least `length` bytes of the file, although additional bytes from the file may be mapped.
+// Ensures that there is one writeable page before address where the file is mapped.
+// Ensures that there is at least `suffix_padding` writable bytes beyond the requested mapping length.
+// madvise()s the file mapping to instruct the kernel to begin reading the data.
+int MmapArray::initMmap(std::string filename, size_t length, size_t suffix_padding) {    
+
+    int fd = open(filename.c_str(), O_RDONLY);
+    if (fd == -1) {
+        return -1;
+    }
+
+    file_mmap_length = round_up_to_page(length + suffix_padding);
+    mmap_length = file_mmap_length + PAGE;
+
+    // allocate the size we need, plus a preceding page
+    void* addr1 = mmap(NULL, mmap_length, PROT_READ|PROT_WRITE, MAP_ANONYMOUS|MAP_PRIVATE|MAP_NORESERVE, -1, 0);
+
+    if (addr1 == (void*) -1) {
+        std::cerr << "reference mmap failed: fd: " << fd;
```

Comment:
Would be useful to get the actual error message here and other places.
```suggestion
        int err = errno;  // Cache this before cerr has a chance to clobber it.
        std::cerr << "reference mmap failed: " << strerror(err) << " on fd: " << fd;
```